### PR TITLE
http: optimize the UnionString append validation

### PR DESCRIPTION
**Pull request by botengyao**
Signed-off-by: Boteng Yao <boteng@google.com>

#23341 refactored http HeaderString to common usage before. This PR is going to add back the optimization to only valid the input (n) rather than the whole appended string buffer each time (n^2). Also added one timeout fuzzy corpus.

Commit Message:  optimize the UnionString validation in append method
Additional Description:
Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: envoy/common/union_string.h [hunk 1]
```diff
@@ -256,6 +257,8 @@ template <class Validator> class UnionStringBase {
 
   bool valid() const { return Validator()(getStringView()); }
 
+  bool validInput(absl::string_view data) const { return Validator()(data); }
+
```
Comment (wbpcode) [inline envoy/common/union_string.h:261]: or just 
```
bool valid(absl::optional<absl::string_view> view) { 
  return view.has_value() ? Validator()(view.value()) : Validator()(getStringView());
}
```

Comment (botengyao) [reply envoy/common/union_string.h]: Good point, done.

**Review by wbpcode — APPROVED** [review]
Thanks. LGTM.

**Pull request merged**